File: Binary Search Tree/Javascript/BST.js
//Javascript implementation of a Binary Search Tree using ES6

class Node {
    constructor(data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class BinarySearchTree {
    constructor() {
        this.root = null;
    }

    insert = data => {
        let newNode = new Node(data);

        this.root === null
            ? (this.root = newNode)
            : this.insertNode(this.root, newNode);
    };

    remove = data => {
        //The root is replaced with the root of the modified tree/
        this.root = this.removeNode(this.root, data);
    };

    inorder = node => {
        if (node !== null) {
            this.inorder(node.left);
            console.log(node.data);
            this.inorder(node.right);
        }
    };

    preorder = node => {
        if (node != null) {
            console.log(node.data);
            this.preorder(node.left);
            this.preorder(node.right);
        }
    };

    postorder = node => {
        if (node != null) {
            this.postorder(node.left);
            this.postorder(node.right);
            console.log(node.data);
        }
    };

  search(node, data) {
    if (node === null)  {
      return null;
    } else if (data < node.data)  {
      return this.search(node.left, data);
    } else if (data > node.data)  {
      return this.search(node.right, data);
    } else {
      return node;
    }
  } 

    insertNode = (node, newNode) => {
        if (newNode.data < node.data) {
            node.left === null
                ? (node.left = newNode)
                : this.insertNode(node.left, newNode);
        } else {
            node.right === null
                ? (node.right = newNode)
                : this.insertNode(node.right, newNode);
        }
    };

    removeNode = (node, target) => {
        if (this.root === null) {
            return null;
        } else if (target < node.data) {
            node.left = this.removeNode(node.left, target);
            return node;
        } else if (target > node.data) {
            node.right = this.removeNode(node.right, target);
        } else {
            if (node.left === null && node.right === null) {
                node = null;
                return node;
            }

            if (node.left === null) {
                node = node.right;
                return node;
            } else if (node.right === null) {
                node = node.left;
                return node;
            }

            let tmpNode = this.findMinNode(node.right);
            node.data = tmpNode.data;

            node.right = this.removeNode(node.right, tmpNode.data);
            return node;
        }
    };

    // Helper functions
    findMinNode = node => {
        if (node.left === null) {
            return node;
        } else {
            this.findMinNode(node.left);
        }
    };
    // getRootNode()
    // inorder(node)
    // preorder(node)
    // postorder(node)
    // search(node, data)
}

File: Binary Search Tree/C++/BST.cpp
#include "BST.h"

using namespace std;

template <class T>
bool addHelper(Node<T> *root, T value)
{
  if (value < root.value)
  {
    if (root.left == NULL)
    {
      root.left = new Node(value);
      return true
    }
    else
    {
      insert(root.left, value);
    }
  }
  else
  {
    if (root.right == NULL)
    {
      root.right = new Node(value);
      return true
    }
    else
    {
      insert(root.right, value);
    }
  }
  return false
}

template <class T>
bool removeHelper(Node<T> *root, T target)
{
  if (root == NULL)
  {
    return false;
  }
  else if (target < root.value)
  {
    root.left = removeHelper(root.left, target)
  }
  else if (target > root.value)
  {
    root.right = removeHelper(root.right, target)
  }
  else if (root.left && root.right)
  {
    temp = findMin(root.right);
    root.data = temp.data;
    root.right = remove(root.data, root.right);
  }
  else
  {
    temp = root;
    if (root.left == NULL)
      root = root.right;
    else if (root.right == NULL)
      root = root.left;
    delete temp;
  }
  return t
}

template <class T>
bool insert(Node<T> *root, T value)
{
  if (root == NULL)
  {
    root = new Node<T>(value) return true
  }
  else
  {
    return insert(root, value)
  }
}

template <class T>
bool remove(Node<T> *root, T value)
{
  if (root == NULL) {
    return false
  } else {
    removeHelper(root, value)
  }
}

template <class T>
Node<T>* findMin(Node<T>* root) {
  if (root == NULL)
    return NULL;
  else if (root.left == NULL)
    return root;
  else
    return findMin(t.left);
}